File: 70-tools/scripts/docs/test_regen_registry.py
"""Tests for regen-registry.py (cycle 57).

The generator walks 90-docs/**/*.md, parses YAML-ish front-matter,
and emits 90-docs/_registry/docs.json. Tests cover:
  - parse_frontmatter happy path (flat keys + list values)
  - parse_frontmatter edge cases (empty list, quoted strings, bools)
  - parse_frontmatter rejects non-front-matter input
  - build_registry shape (version + updated_at + entries)
  - live-repo regression guard (cycle 48 baseline: in sync 659 entries)
"""

from __future__ import annotations

import importlib.util
import json
import subprocess
import sys
from pathlib import Path

import pytest


_SCRIPT = Path(__file__).resolve().parent / "regen-registry.py"


@pytest.fixture(scope="module")
def gen():
    spec = importlib.util.spec_from_file_location("regen_registry", _SCRIPT)
    assert spec is not None and spec.loader is not None
    mod = importlib.util.module_from_spec(spec)
    sys.modules["regen_registry"] = mod
    spec.loader.exec_module(mod)
    return mod


def test_script_exists():
    assert _SCRIPT.exists()


def test_imports_cleanly(gen):
    assert callable(gen.parse_frontmatter)
    assert callable(gen.scan_docs)
    assert callable(gen.build_registry)
    assert callable(gen.main)


def test_parse_frontmatter_minimal(gen):
    """Flat scalars + bool + quoted string parse correctly."""
    text = """---
id: test-1
title: "Test Doc"
status: active
doc_type: adr
topic: test
authoritative: true
last_verified: 2026-05-27
---
body"""
    out = gen.parse_frontmatter(text)
    assert out is not None
    assert out["id"] == "test-1"
    assert out["title"] == "Test Doc"
    assert out["status"] == "active"
    assert out["doc_type"] == "adr"
    assert out["topic"] == "test"
    assert out["authoritative"] is True
    assert out["last_verified"] == "2026-05-27"


def test_parse_frontmatter_lists(gen):
    """YAML list-of-strings parses to Python list."""
    text = """---
id: test-list
title: T
related:
  - foo-1
  - foo-2
authoritative_for:
  - thing-a
  - thing-b
supersedes: []
---"""
    out = gen.parse_frontmatter(text)
    assert out["related"] == ["foo-1", "foo-2"]
    assert out["authoritative_for"] == ["thing-a", "thing-b"]
    assert out["supersedes"] == []


def test_parse_frontmatter_no_yaml_returns_none(gen):
    """Text without leading `---` returns None."""
    assert gen.parse_frontmatter("just body text\nno frontmatter") is None
    assert gen.parse_frontmatter("") is None


def test_parse_frontmatter_unterminated_returns_none(gen):
    """Front-matter that starts with `---` but never closes returns None."""
    text = "---\nid: test\ntitle: never closed"
    assert gen.parse_frontmatter(text) is None


def test_parse_frontmatter_bool_false(gen):
    """`authoritative: false` parses as Python False (not string)."""
    text = """---
id: t
authoritative: false
---"""
    out = gen.parse_frontmatter(text)
    assert out["authoritative"] is False


def test_build_registry_shape(gen):
    """build_registry returns version 2 + updated_at + entries fields."""
    entries = [{"id": "a", "path": "90-docs/a.md", "title": "A"}]
    reg = gen.build_registry(entries)
    assert reg["version"] == 2
    assert "updated_at" in reg
    assert reg["entries"] == entries
    # updated_at is ISO date
    assert len(reg["updated_at"]) == 10  # YYYY-MM-DD


def test_live_repo_in_sync():
    """Live repo registry is in sync per cycle 55 baseline. Regression guard."""
    result = subprocess.run(
        ["python3", str(_SCRIPT), "--check"],
        capture_output=True,
        text=True,
    )
    # exit 0 = in sync; exit 1 = drift
    assert result.returncode == 0, f"registry drift detected: {result.stdout} {result.stderr}"
    assert "in sync" in result.stdout


def test_json_mode_emits_valid_json(gen):
    """`--json` mode produces parseable JSON output."""
    result = subprocess.run(
        ["python3", str(_SCRIPT), "--json"],
        capture_output=True,
        text=True,
    )
    assert result.returncode == 0
    payload = json.loads(result.stdout)
    # Has documented top-level shape
    assert payload["version"] == 2
    assert "updated_at" in payload
    assert "entries" in payload
    assert isinstance(payload["entries"], list)
    # Cycle 55 baseline: 659 entries
    assert len(payload["entries"]) >= 600  # broad lower bound


# ── EDN sidecar (babashka / Clojure consumers) ─────────────────────────────

def test_edn_encoder_primitives(gen):
    """EDN scalar/keyword/collection encoding is correct + injection-safe."""
    assert gen._edn_keyword("doc_type") == ":doc-type"
    assert gen._edn_keyword("authoritative_for") == ":authoritative-for"
    assert gen._edn_value(True) == "true"
    assert gen._edn_value(False) == "false"
    assert gen._edn_value(2) == "2"
    assert gen._edn_value(None) == "nil"
    assert gen._edn_value([]) == "[]"
    assert gen._edn_value(["a", "b"]) == '["a" "b"]'
    # strings: quotes/backslashes/newlines are escaped (no EDN injection)
    assert gen._edn_value('he said "hi"') == '"he said \\"hi\\""'
    assert gen._edn_value("a\\b") == '"a\\\\b"'
    assert gen._edn_value("line1\nline2") == '"line1\\nline2"'
    # nested map uses kebab keyword keys
    assert gen._edn_value({"a_b": 1}) == "{:a-b 1}"


def test_render_edn_shape(gen):
    """render_edn emits a top-level map with kebab keyword keys, one entry per line."""
    reg = gen.build_registry([
        {"path": "90-docs/a.md", "id": "a", "title": 'A "x"', "doc_type": "adr",
         "authoritative": True, "authoritative_for": ["x"]},
    ])
    out = gen.render_edn(reg)
    assert out.startswith("{:version 2\n :updated-at ")
    assert " :entries\n [{" in out
    assert ":doc-type \"adr\"" in out
    assert ":doc_type" not in out          # snake_case never leaks into EDN
    assert out.endswith("]}\n")


def test_edn_mode_emits_parseable_edn(gen):
    """`--edn` mode produces the same shape and round-trips via Python's reader-ish checks."""
    result = subprocess.run(
        ["python3", str(_SCRIPT), "--edn"],
        capture_output=True,
        text=True,
    )
    assert result.returncode == 0
    out = result.stdout
    assert out.startswith("{:version 2\n :updated-at ")
    assert " :entries\n [" in out
    assert ":doc_type" not in out          # all keys kebab-cased
    # balanced delimiters (cheap structural sanity)
    assert out.count("{") == out.count("}")
    assert out.count("[") == out.count("]")

File: 70-tools/scripts/docs/test_regen_registry.clj
;; Tests for regen-registry.clj (babashka port of regen-registry.py).
;; Run: nbb scripts/run-task.cljs test:docs-registry
;; (which is `clojure -Sdeps … -M 70-tools/scripts/docs/test_regen_registry.clj`, from the repo
;; root -- `here` is derived from *file*, so the working directory matters.)
;; Covers parse-frontmatter (happy + edge), entry build/order, JSON+EDN
;; encoders, byte-identity round-trip, and the live-repo freshness guard.

(ns test-regen-registry
  (:require [clojure.test :refer [deftest is run-tests]]
            [babashka.fs :as fs]
            [babashka.process :as p]
            [cheshire.core :as json]
            [clojure.string :as str]))

(def here (fs/parent (fs/absolutize *file*)))
(def impl (str (fs/path here "regen-registry.clj")))

;; Load the implementation for its functions only. Declaring that explicitly is what keeps its
;; entry guard from running -main and exiting this process -- see the guard's own comment for why
;; the default is the other way round.
(System/setProperty "regen-registry.library-load" "1")
(load-file impl)

(defn- pf [t] (regen-registry/parse-frontmatter t))

(deftest script-exists
  (is (fs/exists? impl)))

(deftest parse-frontmatter-minimal
  (let [out (pf "---\nid: x\ntitle: Hello\nauthoritative: true\n---\nbody")]
    (is (= "x" (get out "id")))
    (is (= "Hello" (get out "title")))
    (is (true? (get out "authoritative")))))

(deftest parse-frontmatter-lists
  (let [out (pf "---\nid: x\nrelated:\n  - a\n  - \"b\"\nempty: []\n---\n")]
    (is (= ["a" "b"] (get out "related")))
    (is (= [] (get out "empty")))))

(deftest parse-frontmatter-no-yaml-returns-nil
  (is (nil? (pf "no front matter here"))))

(deftest parse-frontmatter-unterminated-returns-nil
  (is (nil? (pf "---\nid: x\ntitle: y\n"))))

(deftest parse-frontmatter-bool-false
  (is (false? (get (pf "---\nid: x\nauthoritative: false\n---\n") "authoritative"))))

(deftest parse-frontmatter-quoted-arraylike-stays-string
  ;; the tiny parser treats a YAML inline-array value as a scalar string
  (let [out (pf "---\nid: x\nsuperseded_by: [\"2606161200 (D2 only)\"]\n---\n")]
    (is (= "[\"2606161200 (D2 only)\"]" (get out "superseded_by")))))

(deftest build-entry-shape-and-order
  (let [e (regen-registry/build-entry "90-docs/a.md" {"id" "a" "title" "A" "doc_type" "adr"})]
    (is (= "path" (ffirst e)))                       ; path always first
    (is (= ["id" "a"] (second e)))                   ; then surfaced order
    (is (some #(= ["authoritative" false] %) e))     ; authoritative defaulted
    (is (some #(= ["related" []] %) e))))            ; list keys defaulted

(deftest json-encoder-fidelity
  (is (= "true" (#'regen-registry/json-encode true 0)))
  (is (= "\"he said \\\"hi\\\"\"" (#'regen-registry/json-encode "he said \"hi\"" 0)))
  (is (= "[]" (#'regen-registry/json-encode [] 0))))

(deftest edn-encoder-fidelity
  (is (= ":doc-type" (#'regen-registry/edn-keyword "doc_type")))
  (is (= "\"a\\\\b\"" (#'regen-registry/edn-value "a\\b")))
  (is (= "[\"a\" \"b\"]" (#'regen-registry/edn-value ["a" "b"]))))

(deftest render-json-shape
  (let [out (regen-registry/render-json
             [(regen-registry/build-entry "90-docs/a.md" {"id" "a" "title" "A"})] "2026-06-16")
        parsed (json/parse-string out false)]
    (is (str/starts-with? out "{\n  \"version\": 2,\n"))
    (is (= 2 (get parsed "version")))
    (is (= 1 (count (get parsed "entries"))))
    (is (str/ends-with? out "}\n"))))

(deftest render-edn-shape
  (let [out (regen-registry/render-edn
             [(regen-registry/build-entry "90-docs/a.md" {"id" "a" "doc_type" "adr"})] "2026-06-16")]
    (is (str/starts-with? out "{:version 2\n :updated-at "))
    (is (str/includes? out ":doc-type \"adr\""))
    (is (not (str/includes? out ":doc_type")))      ; snake never leaks into EDN
    (is (str/ends-with? out "]}\n"))))

(deftest live-repo-in-sync
  ;; The committed registry must match what we'd regenerate (json + edn).
  ;;
  ;; This used to shell out to `bb 70-tools/scripts/docs/regen-registry.clj --check`. Two things
  ;; were wrong with that after ADR-2607173000 retired babashka: the binary is gone on any fleet
  ;; node, and the subprocess bought nothing -- the implementation is ALREADY load-file'd into this
  ;; process above, and its -main RETURNS an exit code rather than calling System/exit (the exit is
  ;; in the script's own entry guard). So call it directly and capture what it prints. One fewer
  ;; process, one fewer runtime assumption, and a failure now reports in this process's output
  ;; instead of inside a captured string.
  ;; Capture *err* as well as *out*: -main prints the "in sync" line to stdout but every drift
  ;; message to stderr, so capturing only stdout reports a failure with an empty explanation.
  (let [out (java.io.StringWriter.)
        err (java.io.StringWriter.)
        exit (binding [*out* out *err* err] (regen-registry/-main "--check"))
        text (str out err)]
    (is (zero? exit) (str "registry drift: " text))
    (is (str/includes? text "in sync"))))

;; Unconditional exit. The guard was `(when (= *file* (System/getProperty "babashka.file")) …)`,
;; which is false under every runtime except babashka -- so run under anything else this file ran
;; its tests, printed the failures, and then exited 0 regardless. A test runner that always
;; succeeds does not report a green suite; it reports nothing, in the shape of a green suite.
(let [{:keys [fail error]} (run-tests 'test-regen-registry)]
  (System/exit (if (pos? (+ fail error)) 1 0)))
